src/rewrite/vct/rewrite/VariableToPointer.scala

```diff
@@ -0,0 +1,104 @@
+package vct.rewrite
+
+import vct.col.ast._
+import vct.col.ref._
+import vct.col.origin._
+import vct.col.rewrite.{Generation, Rewriter, RewriterBuilder, Rewritten}
+import vct.col.util.AstBuildHelpers._
+import vct.col.util.SuccessionMap
+import vct.result.VerificationError.UserError
+
+import scala.collection.mutable
+
+case object VariableToPointer extends RewriterBuilder {
+  override def key: String = "variableToPointer"
+
+  override def desc: String = "Translate every local and field to a pointer such that it can have its address taken"
+
+  case class UnsupportedAddrOf(loc: Expr[_]) extends UserError {
+    override def code: String = "unsupportedAddrOf"
+
+    override def text: String = loc.o.messageInContext("Taking an address of this expression is not supported")
+  }
+}
+
+case class VariableToPointer[Pre <: Generation]() extends Rewriter[Pre] {
+
+  import VariableToPointer._
+
+  var stageTwo = false
+  val addressedSet: mutable.Set[Node[Pre]] = new mutable.HashSet[Node[Pre]]()
+  val heapVariableMap: SuccessionMap[HeapVariable[Pre], HeapVariable[Post]] = SuccessionMap()
+  val variableMap: SuccessionMap[Variable[Pre], Variable[Post]] = SuccessionMap()
+  val fieldMap: SuccessionMap[InstanceField[Pre], InstanceField[Post]] = SuccessionMap()
+
+  override def dispatch(program: Program[Pre]): Program[Rewritten[Pre]] = {
+    super.dispatch(program)
+    stageTwo = true
```

Comment:
Instead of two stages, maybe you can do:

`program.collect({case a :AddrOf[Pre] =>a})`
to collect all adrr offs and process them separately.

Reply:
Ah I didn't know that existed